### DISCUSSION TITLE
Union type auto reduction/inference on pattern matches.

### DISCUSSION THREAD

**Discussion by Swoorup** [scala/scala3#12810] [General Discussion]
In typescript I can do like the following: 

```typescript
type U = "A" | "B" | "C"
function uBut(x: U): "B" | "C" {
  switch (x) {
    case "A":
      return "B"
    default:
      return x
  }
}
```

Here, `x` is deduced to a union of `"B" | "C"`. 

With scala 3 union unfortunately you can't do so at the moment. The following fails to compile:

```scala
type U = Int | String | Boolean

def uButString(u: U): Int | Boolean =
  u match
    case s: String => 0
    case x => x
```

This will lead people to instead use `asInstanceOf` which is bad imo. You can still make it typesafe if you annotate the last type, without warnings like

```scala
type U = Int | String | Boolean

def uButString(u: U): Int | Boolean =
  u match
    case s: String => 0
    case x: (Int | Boolean) => x
```

But it would be nice, if scala was able to auto deduce the type of `x` to `Int | Boolean`, leading to more natural code and less usage of `asInstanceOf`.

If there was a condition on the pattern match, we could retain the current behavior as is.

**Comment by Swoorup** [comment]
This can be archived, as I created a feature request on dotty

**Reply by dwijnand** [reply]
https://github.com/lampepfl/dotty-feature-requests/issues/223
